pkg/ipcache/metadata.go

```diff
@@ -125,19 +125,27 @@ func (m *metadata) upsertLocked(prefix netip.Prefix, src source.Source, resource
 	m.m[prefix].logConflicts(log.WithField(logfields.CIDR, prefix))
 }
 
-// GetMetadataLabelsByIP returns the associated labels with an IP. The caller must
-// not modifying the returned object as it's a live reference to the underlying
-// map.
+// GetMetadataLabelsByIP returns the associated labels with an IP.
 func (ipc *IPCache) GetMetadataLabelsByIP(addr netip.Addr) labels.Labels {
 	prefix := netip.PrefixFrom(addr, addr.BitLen())
-	ipc.metadata.RLock()
-	defer ipc.metadata.RUnlock()
-	if info := ipc.metadata.getLocked(prefix); info != nil {
+	if info := ipc.GetMetadataByPrefix(prefix); info != nil {
 		return info.ToLabels()
 	}
 	return nil
 }
 
+// GetMetadataByPrefix returns full metadata for a given IP as a copy.
+func (ipc *IPCache) GetMetadataByPrefix(prefix netip.Prefix) prefixInfo {
+	ipc.metadata.RLock()
+	defer ipc.metadata.RUnlock()
+	m := ipc.metadata.getLocked(prefix)
+	n := make(prefixInfo, len(m))
+	for k, v := range m {
+		n[k] = v
```

Comment:
`v` is a pointer of type `*resourceInfo` which can be mutated in place e.g. here:

https://github.com/cilium/cilium/blob/1c01d518141b011ab75f398a350b4ead7c6b316b/pkg/ipcache/metadata.go#L121

I think we need to create a copy of `v` here, otherwise we can have mutation happen on it while we are reading from it - or alternatively only return the `Source()` part here.

Reply:
Good catch, addressed

<details><summary>Details</summary>
<p>

```diff
commit 446e180d3e9e58336b5425e3bb0a056a227e1083
Author: Chris Tarazi <chris@isovalent.com>
Date:   Wed Apr 12 16:04:48 2023 -0700

    fixup! ipcache: Define new GetMetadataByPrefix
    
    Signed-off-by: Chris Tarazi <chris@isovalent.com>

diff --git a/pkg/ipcache/metadata.go b/pkg/ipcache/metadata.go
index c631d9f433..84dfd37112 100644
--- a/pkg/ipcache/metadata.go
+++ b/pkg/ipcache/metadata.go
@@ -141,7 +141,7 @@ func (ipc *IPCache) GetMetadataByPrefix(prefix netip.Prefix) PrefixInfo {
 	m := ipc.metadata.getLocked(prefix)
 	n := make(PrefixInfo, len(m))
 	for k, v := range m {
-		n[k] = v
+		n[k] = v.DeepCopy()
 	}
 	return n
 }
diff --git a/pkg/ipcache/types.go b/pkg/ipcache/types.go
index fc7d7709c7..f31385962c 100644
--- a/pkg/ipcache/types.go
+++ b/pkg/ipcache/types.go
@@ -21,6 +21,9 @@ import (
 // independently based on the ResourceID of the origin of that information, and
 // provides convenient accessors to consistently merge the stored information
 // to generate ipcache output based on a range of inputs.
+//
+// Note that when making a copy of this object, resourceInfo is pointer which
+// means it needs to be deep-copied via (*resourceInfo).DeepCopy().
 type PrefixInfo map[types.ResourceID]*resourceInfo
 
 // IdentityOverride can be used to override the identity of a given prefix.
@@ -105,6 +108,16 @@ func (m *resourceInfo) isValid() bool {
 	return false
 }
 
+func (m *resourceInfo) DeepCopy() *resourceInfo {
+	n := new(resourceInfo)
+	n.labels = labels.NewFrom(m.labels)
+	n.source = m.source
+	n.identityOverride = m.identityOverride
+	n.tunnelPeer = m.tunnelPeer
+	n.encryptKey = m.encryptKey
+	return n
+}
+
 func (s PrefixInfo) isValid() bool {
 	for _, v := range s {
 		if v.isValid() {
```

</p>
</details>